### PR TITLE
Enhance filenames with collection context when downloading datasets from GUI.

### LINKED ISSUE #2140: Feature request - Filenames based on dataset collection identifier
I'm trying to incorporate the use of dataset collections to simplify workflows, esp. ones with many samples (e.g. RNA-Seq). One issue I'm running into is when I need to get the data outside of galaxy, either for external analysis (e.g. in R) or visualization on UCSC. In both cases, the data is referenced as the "name" of the file (e.g. "Tophat on dataset 1 and 2: accepted hits") instead of the collection id (e.g. sample A). It would be nice if there was a way to expose this "collection id" when exporting data.

### PR DESCRIPTION
**Pull request by jmchilton**
- Reduce duplication related to calculating filename.
- Formalized pattern for normal dataset download using string.Tempalte - "``Galaxy${hid}-[${name}].${ext}``".
- In the context of collections, hid and name of the datasets are pretty worthless so I used a second pattern for collections based on where it is in the current GUI frame. The new pattern is "``Galaxy${hdca_hid}-[${hdca_name}__${element_identifier}].${ext}``".
- The patterns make this a little more tweakable - for people who want something different it should be pretty straight forward to make these patterns configurable.

Fixes #2140 I think.

### CODE REVIEW

**Comment by jmchilton** [comment]
Thanks for the merge @martenson - glad this made it into 17.05.

File: lib/galaxy/datatypes/data.py [hunk 1]
```diff
@@ -355,11 +359,8 @@ def display_data(self, trans, data, preview=False, filename=None, to_ext=None, *
                 return self._archive_composite_dataset( trans, data, **kwd )
             else:
                 trans.response.headers['Content-Length'] = int( os.stat( data.file_name ).st_size )
-                if not to_ext:
-                    to_ext = data.extension
-                fname = ''.join(c in FILENAME_VALID_CHARS and c or '_' for c in data.name)[0:150]
-                trans.response.set_content_type( "application/octet-stream" )  # force octet-stream so Safari doesn't append mime extensions to filename
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/data.py:361]: @jmchilton Was this line removed on purpose?

Comment (jmchilton) [reply lib/galaxy/datatypes/data.py:361]: Nope :(

Comment (jmchilton) [reply lib/galaxy/datatypes/data.py:361]: https://github.com/galaxyproject/galaxy/pull/4004 - sorry about that and thanks for the catch.

**Comment by mvdbeek** [comment]
This is not in 17.05 though, what's the procedure now, can I push this to the 17.05 branch?

**Comment by mvdbeek** [comment]
Actually it is and it works fine :D

**Comment by martenson** [comment]
@mvdbeek yeah, this has been merged pre-freeze :)

**Pull request merged**